examples/vision/video_transformers.py

```diff
@@ -0,0 +1,408 @@
+
+
+"""
+Now, we can create a subclassed layer for the Transformer. 
+"""
+
+
+class TransformerEncoder(layers.Layer):
+    def __init__(self, embed_dim, dense_dim, num_heads, **kwargs):
+        super().__init__(**kwargs)
+        self.embed_dim = embed_dim
+        self.dense_dim = dense_dim
+        self.num_heads = num_heads
+        self.attention = layers.MultiHeadAttention(
+            num_heads=num_heads, key_dim=embed_dim, dropout=0.3
+        )
+        self.dense_proj = keras.Sequential(
+            [layers.Dense(dense_dim, activation=tf.nn.gelu), layers.Dense(embed_dim),]
+        )
+        self.layernorm_1 = layers.LayerNormalization()
+        self.layernorm_2 = layers.LayerNormalization()
+
+    def call(self, inputs, mask=None):
+        if mask is not None:
+            mask = mask[:, tf.newaxis, :]
+        attention_output = self.attention(inputs, inputs, attention_mask=mask)
+        proj_input = self.layernorm_1(inputs + attention_output)
+        proj_output = self.dense_proj(proj_input)
+        return self.layernorm_2(proj_input + proj_output)
+
+    def get_config(self):
+        config = super(TransformerEncoder, self).get_config()
+        config.update(
+            {
+                "embed_dim": self.embed_dim,
+                "num_heads": self.num_heads,
+                "dense_dim": self.dense_dim,
+            }
+        )
+        return config
+
+
+"""
+## Utility functions for training
+"""
+
+
+def get_compiled_model():
+    sequence_length = MAX_SEQ_LENGTH
+    embed_dim = NUM_FEATURES
+    dense_dim = 4
+    num_heads = 1
+    classes = len(label_processor.get_vocabulary())
+
+    inputs = keras.Input(shape=(None, None))
+    x = PositionalEmbedding(
+        sequence_length, embed_dim, name="frame_position_emebedding"
+    )(inputs)
+    x = TransformerEncoder(embed_dim, dense_dim, num_heads, name="transformer_layer")(
+        x, mask=None
+    )
+    x = layers.GlobalMaxPooling1D()(x)
+    x = layers.Dropout(0.5)(x)
+    outputs = layers.Dense(classes, activation="softmax")(x)
+    model = keras.Model(inputs, outputs)
+
+    model.compile(
+        optimizer="adam", loss="sparse_categorical_crossentropy", metrics=["accuracy"]
+    )
+    return model
+
+
+def run_experiment():
+    filepath = "/tmp/video_classifier"
+    checkpoint = keras.callbacks.ModelCheckpoint(
+        filepath, save_weights_only=True, save_best_only=True, verbose=1
+    )
+
+    model = get_compiled_model()
+    history = model.fit(
+        train_data,
+        train_labels,
+        validation_split=0.15,
+        epochs=EPOCHS,
+        callbacks=[checkpoint],
+    )
+
+    model.load_weights(filepath)
+    _, accuracy = model.evaluate(test_data, test_labels)
+    print(f"Test accuracy: {round(accuracy * 100, 2)}%")
+
+    return model
+
+
+"""
+## Model training and inference
+"""
+
+trained_model = run_experiment()
+
+"""
+**Note**: This model has ~4.23 Million parameters which is way more than the sequence
```

Comment:
Could we make this number a lot lower?

Reply:
I had this thought when I called `summary()` 😂  But if you see I have kept the hyperparameters to a bare minimum:

* `num_heads`: 1
* `dense_dim`: 4

To speed up computation and to reduce the parameters further I used a DenseNet121 which yields a tensor of shape `(bs, 1024)` (with pooling included).